File: ports/go/src/core/global/bcssignal.go
package global

import "sync"

// BcsSignal maps the C++ signal/slot paradigm to Go channels and callbacks.
type BcsSignal struct {
	mu        sync.RWMutex
	listeners []func(interface{})
}

// Connect adds a listener callback to the signal.
func (s *BcsSignal) Connect(callback func(interface{})) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.listeners = append(s.listeners, callback)
}

// Emit broadcasts the payload to all connected listeners.
func (s *BcsSignal) Emit(payload interface{}) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	for _, listener := range s.listeners {
		go listener(payload) // Asynchronous execution default
	}
}

File: ports/rust/src/core/global/bcssignal.rs
use std::sync::{Arc, Mutex};
use std::thread;

pub struct BcsSignal<T: Clone + Send + 'static> {
    listeners: Arc<Mutex<Vec<Box<dyn Fn(&T) + Send + Sync>>>>,
}

impl<T: Clone + Send + 'static> BcsSignal<T> {
    pub fn new() -> Self {
        BcsSignal {
            listeners: Arc::new(Mutex::new(Vec::new())),
        }
    }

    pub fn connect<F>(&self, callback: F)
    where
        F: Fn(&T) + Send + Sync + 'static,
    {
        let mut listeners = self.listeners.lock().unwrap();
        listeners.push(Box::new(callback));
    }

    pub fn emit(&self, payload: &T) {
        let listeners = self.listeners.lock().unwrap();
        for listener in listeners.iter() {
            listener(payload);
        }
    }

    pub fn emit_async(&self, payload: T) {
        let listeners = Arc::clone(&self.listeners);
        thread::spawn(move || {
            let listeners = listeners.lock().unwrap();
            for listener in listeners.iter() {
                listener(&payload);
            }
        });
    }
}
